asf_search/search/search_generator.py

```diff
@@ -78,78 +78,78 @@ def search_generator(
+            subquery_count += len(last_page)
+            last_page.searchComplete = subquery_count == subquery_max_results or total == maxResults
+            yield last_page
+            
+            if last_page.searchComplete:
+                if subquery_count == subquery_max_results and maxResults is None: # end of subquery
+                    cmr_search_after_header = None
+                elif total == maxResults:
+                    opts.session.headers.pop('CMR-Search-After', None)
+                    return
                 else:
-                    count += len(hits)
-                    yield ASFSearchResults(hits, opts=opts)
-
+                    cmr_search_after_header = None
+        
         opts.session.headers.pop('CMR-Search-After', None)
 
 
 @retry(reraise=True,
-       retry=retry_if_exception_type((TimeoutError, ASFSearch5xxError)),
-       wait=wait_exponential(multiplier=1, min=4, max=10),
-       stop=stop_after_delay(340),
+       retry=retry_if_exception_type(CMRIncompleteError),
+       wait=wait_fixed(2),
+       stop=stop_after_delay(120),
+    )
+def query_cmr(session: ASFSession, url: str, translated_opts: dict, sub_query_count: int):
+    response = get_page(session=session, url=url, translated_opts=translated_opts)
+
+    items = [ASFProduct(f, session=session) for f in response.json()['items']]
+    hits: int = response.json()['hits'] # total count of products given search opts
+
+    # sometimes CMR returns results with the wrong page size
+    if len(items) != INTERNAL.CMR_PAGE_SIZE and len(items) + sub_query_count < hits:
+        raise CMRIncompleteError(f"CMR returned page of incomplete results. Expected {min(INTERNAL.CMR_PAGE_SIZE, hits - sub_query_count)} results, got {len(items)}")
+
+    return items, hits, response.headers.get('CMR-Search-After', None)
+    
+
+def process_page(items: list[ASFProduct], max_results: int, subquery_max_results: int, total: int, subquery_count: int, opts: ASFSearchOptions):
+    if max_results is None:
+        last_page = ASFSearchResults(items[:min(subquery_max_results - subquery_count, len(items))], opts=opts)
+    else:
+        last_page = ASFSearchResults(items[:min(max_results - total, len(items))], opts=opts)
+    total += len(last_page)
+
+    return total, last_page
+
+
+@retry(reraise=True,
+       retry=retry_if_exception_type(ASFSearch5xxError),
+       wait=wait_exponential(multiplier=1, min=3, max=10),  # Wait 2^x * 1 starting with 3 seconds, max 10 seconds between retries
+       stop=stop_after_delay(60),
     )
-def get_page(session: ASFSession, url: str, translated_opts: list, search_opts: ASFSearchOptions) -> Response:
+def get_page(session: ASFSession, url: str, translated_opts: list) -> Response:
     try:
-        response = session.post(url=url, data=translated_opts, timeout=170)
+        response = session.post(url=url, data=translated_opts, timeout=30)
```

Comment:
Does @retry above force a timeout too? If so, we probably could drop the timeout key here completely and only maintain it in one place?

Reply:
Not quite, `wait` in @retry defines how long to wait before retries. `stop=stop_after_delay` acts like a timeout, counting down from the first attempt. 

But I went ahead and changed both  @retry decorators in search_generator.py to use `stop=stop_after_attempt(3)`, so now the only timeout should be defined in the CMR request!